src/quic/streams.cc

```diff
@@ -1,61 +1,1052 @@
+  if (is_destroyed() || state_->read_ended) return;
+  ngtcp2_conn_shutdown_stream_read(session(), id(), error.code());
+  EndReadable();
+}
+
+void Stream::ReceiveStreamReset(uint64_t final_size, QuicError error) {
+  // Importantly, reset stream only impacts the inbound data flow. It has no
+  // impact on the outbound data flow. It is essentially a signal that the peer
+  // has abruptly terminated the writable end of their stream with an error.
+  // Any data we have received up to this point remains in the queue waiting to
+  // be read.
+  EndReadable(final_size);
+  EmitReset(error);
+}
+
+// ============================================================================
+
+void Stream::EmitBlocked() {
+  // state_->wants_block will be set from the javascript side if the
+  // stream object has a handler for the blocked event.
+  if (is_destroyed() || !env()->can_call_into_js() ||
+      state_->wants_block == 0) {
+    return;
+  }
+  CallbackScope<Stream> cb_scope(this);
+  MakeCallback(BindingData::Get(env()).stream_blocked_callback(), 0, nullptr);
+}
+
+void Stream::EmitClose(const QuicError& error) {
+  if (is_destroyed() || !env()->can_call_into_js()) return;
+  CallbackScope<Stream> cb_scope(this);
+  Local<Value> err;
+  if (!error.ToV8Value(env()).ToLocal(&err)) return;
 
-void Stream::Schedule(Stream::Queue* queue) {}
-void Stream::Unschedule() {}
+  MakeCallback(BindingData::Get(env()).stream_close_callback(), 1, &err);
+}
+
+void Stream::EmitHeaders() {
+  if (is_destroyed() || !env()->can_call_into_js() ||
+      state_->wants_headers == 0) {
+    return;
+  }
+  CallbackScope<Stream> cb_scope(this);
+
+  Local<Value> argv[] = {
+      Array::New(env()->isolate(), headers_.data(), headers_.size()),
+      Integer::NewFromUnsigned(env()->isolate(),
+                               static_cast<uint32_t>(headers_kind_))};
+
+  headers_.clear();
+
+  MakeCallback(
+      BindingData::Get(env()).stream_headers_callback(), arraysize(argv), argv);
+}
+
+void Stream::EmitReset(const QuicError& error) {
+  if (is_destroyed() || !env()->can_call_into_js() ||
+      state_->wants_reset == 0) {
+    return;
+  }
+  CallbackScope<Stream> cb_scope(this);
+  Local<Value> err;
+  if (!error.ToV8Value(env()).ToLocal(&err)) return;
+
+  MakeCallback(BindingData::Get(env()).stream_reset_callback(), 1, &err);
+}
+
+void Stream::EmitWantTrailers() {
+  if (is_destroyed() || !env()->can_call_into_js() ||
+      state_->wants_trailers == 0) {
+    return;
+  }
+  CallbackScope<Stream> cb_scope(this);
+  MakeCallback(BindingData::Get(env()).stream_trailers_callback(), 0, nullptr);
+}
+
+// ============================================================================
+
+void Stream::Schedule(Stream::Queue* queue) {
+  if (is_destroyed() || outbound_ == nullptr) return;
+  // If this stream is not already in the queue to send data, add it.
+  if (stream_queue_.IsEmpty()) queue->PushBack(this);
```

Comment:
The comment and the implementation differs. Is this intended?

Reply:
Not sure what you mean